nixos/modules/services/x11/redshift.nix

```diff
@@ -91,48 +63,77 @@ in {
       '';
     };
 
-    extraOptions = mkOption {
-      type = types.listOf types.str;
-      default = [];
-      example = [ "-v" "-m randr" ];
+    settings = mkOption {
+      type = types.submodule {
+        freeformType = with types; (attrsOf (attrsOf (oneOf [str int float])));
```

Comment:
Not adding `bool` here since the only bool option in documentation is `fade`, and it would need to be converted from `true`/`false` to `1`/`0`.

Reply:
Maybe `settingsFormat.types` already does this conversion :thinking: ?

Reply:
But `settingsFormat.types` seems to support `null` too, that I don't think it is desirable here.

Reply:
One of the main reasons these formats have been abstracted into `pkgs.formats` is that you won't have to encode the type of them yourself. INI supports booleans, integers and nulls, but there's no need to worry about that since all the valid types and their string generation is handled by the `generator` function, which matches what its `type` can represent. So I suggest just using `settingsFormat.generate` here. This way the module can be reused for future redshift versions, even if that newly uses booleans and nulls in its config.

Note however that redshift uses integers 0/1 to encode `fade`, not booleans. So even though the Nix `type` does allow booleans you'll still have to use 0/1 to encode `fade`.

Reply:
@Infinisil So I remove the `type` here or change it to `type = types.submodule { freeFormType = settingsFormat.types; }` like I saw in RFC42 :thinking: ?

Asking because I am already using `settingsFormat.generate` (I may be using it wrong though): https://github.com/NixOS/nixpkgs/pull/108739/files#diff-61e7d28082c9dd1fe9bd0282f55b1cd33f1397fa87b424b6fdf71253e591bc02R122

Reply:
BTW, the reason I decided to overwrite the types is to avoid some typos in the config, like the mentioned `fade` case already or someone trying to pass something as `null` to remove a default config (AFAIK this will not happen). And `redshift` doesn't have a release for almost 3 years already, so I think they breaking their config now is really unlikely.

But I think it is fair enough to use the more generic types too.

Reply:
If the tests implemented now it was easy to check what works and what doesn't.